pkg/server/restart.go

```diff
@@ -179,140 +179,130 @@ func (c *criService) loadContainer(ctx context.Context, cntr containerd.Containe
+					return errors.Wrap(err, "failed to create container io")
+				}
 			case runtime.ContainerState_CONTAINER_RUNNING:
+				// Container was in running state, but its task has been deleted,
+				// set unknown exited state. Container io is not needed in this case.
+				status.FinishedAt = time.Now().UnixNano()
+				status.ExitCode = unknownExitCode
+				status.Reason = unknownExitReason
 			default:
-				// This may happen if containerd gets restarted after task is started, but
-				// before status is checkpointed.
-				status.StartedAt = time.Now().UnixNano()
-				status.Pid = t.Pid()
+				// Container is in exited/unknown state, return the status as it is.
 			}
-		case containerd.Stopped:
-			// Task is stopped. Updata status and delete the task.
-			if _, err := t.Delete(ctx, containerd.WithProcessKill); err != nil && !errdefs.IsNotFound(err) {
-				return container, errors.Wrap(err, "failed to delete task")
+		} else {
+			// Task status is found. Update container status based on the up-to-date task status.
+			switch s.Status {
+			case containerd.Created:
+				// Task has been created, but not started yet. This could only happen if containerd
+				// gets restarted during container start.
+				// Container must be in `CREATED` state.
+				if _, err := t.Delete(ctx, containerd.WithProcessKill); err != nil && !errdefs.IsNotFound(err) {
+					return errors.Wrap(err, "failed to delete task")
+				}
+				if status.State() != runtime.ContainerState_CONTAINER_CREATED {
+					return errors.Errorf("unexpected container state for created task: %q", status.State())
+				}
+			case containerd.Running:
+				// Task is running. Container must be in `RUNNING` state, based on our assuption that
+				// "task should not be started when containerd is down".
+				switch status.State() {
+				case runtime.ContainerState_CONTAINER_EXITED:
+					return errors.Errorf("unexpected container state for running task: %q", status.State())
+				case runtime.ContainerState_CONTAINER_RUNNING:
+				default:
+					// This may happen if containerd gets restarted after task is started, but
+					// before status is checkpointed.
+					status.StartedAt = time.Now().UnixNano()
+					status.Pid = t.Pid()
+				}
+			case containerd.Stopped:
+				// Task is stopped. Updata status and delete the task.
+				if _, err := t.Delete(ctx, containerd.WithProcessKill); err != nil && !errdefs.IsNotFound(err) {
+					return errors.Wrap(err, "failed to delete task")
+				}
+				status.FinishedAt = s.ExitTime.UnixNano()
+				status.ExitCode = int32(s.ExitStatus)
+			default:
+				return errors.Errorf("unexpected task status %q", s.Status)
 			}
-			status.FinishedAt = s.ExitTime.UnixNano()
-			status.ExitCode = int32(s.ExitStatus)
-		default:
-			return container, errors.Errorf("unexpected task status %q", s.Status)
 		}
+		return nil
+	}()
+	if err != nil {
+		logrus.WithError(err).Errorf("Failed to load container status for %q", id)
+		status = unknownContainerStatus()
```

Comment:
if i'm reading this right the logic has changed here.. in the prior implementation the errors would return early.. now they get logged but we do not return err early?

Reply:
Yeah, because we want to continue loading the container/sandbox with status unknown.